core/arch/aarch64/emit_utils.c

```diff
@@ -124,11 +136,28 @@ insert_exit_stub_other_flags(dcontext_t *dcontext, fragment_t *f, linkstub_t *l,
         /* mov x0, ... */
         pc = insert_mov_imm(pc, DR_REG_X0, (ptr_int_t)l);
         num_nops_needed = 4 - (pc - write_stub_pc - 1);
-        /* ldr x1, [x(stolen), #(offs)] */
-        *pc++ = (0xf9400000 | 1 | (dr_reg_stolen - DR_REG_X0) << 5 |
-                 get_fcache_return_tls_offs(dcontext, f->flags) >> 3 << 10);
+        uint target_pc_slot_offs = (uint *)get_target_pc_slot(f, stub_pc) - pc;
+        /* ldr x1, [pc, target_pc_slot_offs * AARCH64_INSTR_SIZE] */
+        *pc++ = (0x58000000 | (DR_REG_X1 - DR_REG_X0) | target_pc_slot_offs << 5);
         /* br x1 */
         *pc++ = BR_X1_INST;
+        /* Fill up with NOPs, depending on how many instructions we needed to move
+         * the immediate in a register. Ideally we would skip adding NOPs, but
+         * lots of places expect the stub size to be fixed.
+         */
+        for (uint j = 0; j < num_nops_needed; j++)
+            *pc++ = NOP_INST;
+        /* The final slot is a data slot, which will hold the address of either
+         * the fcache-return routine or the linked fragment.
+         */
+        ASSERT(sizeof(app_pc) == 8);
+        pc += DIRECT_EXIT_STUB_DATA_SZ / sizeof(uint);
+        /* We start off with the fcache-return routine address in the slot. */
```

Comment:
This works under the assumption that the `fcache-return` routine is shared between all threads. @derekbruening Would you consider that okay to assume? If not, the alternative is Option 4a at https://github.com/DynamoRIO/dynamorio/wiki/Linking-Far-Fragments-on-AArch64.

Reply:
> This works under the assumption that the fcache-return routine is shared between all threads. @derekbruening Would you consider that okay to assume?

In general the "gencode" comes in both thread-private and thread-shared varieties.  But any one fragment, and its stubs, will only ever use one or the other, depending on whether the fragment is private or shared.  (Plus, IIRC 64-bit for both x86 and aarch doesn't have private gencode at all anyway?)

Reply:
Yes AArch64 has shared fragments (hence all the consistency related issues too). I added an ASSERT as discussed, to verify that the `fcache-return` routine from thread dcontext is same as the one from `GLOBAL_DCONTEXT`. This seems a strong assumption for this design, so I think it'll be good to add asserts.

Reply:
> Yes AArch64 has shared fragments

Note that "gencode" in DR refers to just the synthetic code pointed at by generated_code_t (so fcache_{enter,return} and the IBL and variants), not to fragments.  X64 (so x86-64 and A64)'s private fragments use shared gencode.

Reply:
Yes, got it. Thanks.